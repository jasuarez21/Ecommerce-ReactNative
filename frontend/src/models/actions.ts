import { Moto } from "./moto";
import { User } from "./user";

export type MotoActions = 
    | {type: 'GET_MOTOS'; data: Array<Moto>}
    | {type: 'ADD_TO_CART'; moto: Moto}
    | {type: 'LOAD_CART';}
    | {type: 'DELETE_CART'; moto: Moto}
    | {type: 'LOG_IN'; data: User}
    | {type: 'LOG_OUT'; data: User}
    | {type: 'SIGN_UP'; data: User}
    | {type: 'UPDATE_CART'}
    | {type: 'DELETE_ACCOUNT'}
    | {type: 'ERROR'}