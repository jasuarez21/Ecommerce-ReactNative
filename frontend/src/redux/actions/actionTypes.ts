const actionTypes = {
  GET_MOTOS: 'GET_MOTOS',
  LOG_IN: 'LOG_IN',
  LOG_OUT: 'LOG_OUT',
  SIGN_UP: 'SIGN_UP',
  ADD_TO_CART: 'ADD_TO_CART',
  LOAD_CART: 'LOAD_CART',
  DELETE_CART: 'DELETE_CART',
  UPDATE_CART: 'UPDATE_CART',
  SAVE_PRODUCT: 'SAVE_PRODUCT',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  ERROR: 'ERROR'
};

export default actionTypes;
