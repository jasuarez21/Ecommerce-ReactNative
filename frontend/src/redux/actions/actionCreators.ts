import { User } from './../../models/user';
import axios from 'axios';
import {URL_MOTOS, URL_LOGIN, URL_LOGOUT, URL_SIGNUP, URL_DELETE_ACCOUNT, URL_SAVE_PRODUCTS} from '@env';
import actionTypes from './actionTypes';
import {Moto} from '../../models/moto';
import {Dispatch} from 'redux';
import {MotoActions} from '../../models/actions';
import { Order } from '../../models/order';

export function GetMotos() {
  return async (dispatch: Dispatch) => {
      try {   
      const {data} = await axios.get(URL_MOTOS);
      dispatch({
        type: actionTypes.GET_MOTOS,
        data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}



export function login(email: string, password: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGIN, {email, password});
        dispatch({
          type: actionTypes.LOG_IN,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function logout() {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGOUT);
        dispatch({
          type: actionTypes.LOG_OUT,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}


export function signUp(email: string, password: string, alias: string, avatar: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_SIGNUP, {email, password, alias, avatar});
        dispatch({
          type: actionTypes.SIGN_UP,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function addToCart(moto: Moto): MotoActions {
  return {
    type: 'ADD_TO_CART',
    moto
  };
}

export function loadCart(): MotoActions {
  return {
    type: 'LOAD_CART'
  };
}

export function deleteFromCart(moto: Moto): MotoActions {
  return {
    type: 'DELETE_CART',
    moto,
  };
}

export function reloadCart(moto: Array<Moto>){
  return async (dispatch: Dispatch) => {
    try {   
      for(let i = 0; i < moto.length; i++){
        await axios.post(`${URL_MOTOS}/${moto[i]._id}`, moto[i]);
        dispatch({
          type: actionTypes.UPDATE_CART
        });
      }
  } catch (error) {
    dispatch({
      type: actionTypes.ERROR,
    });
  }
}
}

export function saveOrder(user: User, cart: Order, token: string){
  user.cart.push(cart);
  const config = { 
    headers: {
        authorization: `Bearer ${token}`,
    }
  }
    return async (dispatch: Dispatch) => {
      try {   
          await axios.post(URL_SAVE_PRODUCTS, {_id: user._id, cart: user.cart}, config);
          dispatch({
            type: actionTypes.SAVE_PRODUCT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function deleteAccount(user: User, token: string){
    return async (dispatch: Dispatch) => {
      try {   
          await axios.delete(URL_DELETE_ACCOUNT, {headers: {
              authorization: `Bearer ${token}`,
          },
              data:{_id: user._id}
          });
          dispatch({
            type: actionTypes.DELETE_ACCOUNT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}