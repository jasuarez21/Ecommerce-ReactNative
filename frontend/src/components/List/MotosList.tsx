/* eslint-disable react-hooks/exhaustive-deps */
/* eslint-disable no-underscore-dangle */
import React, {useEffect} from 'react';
import {RootState} from '../../redux/store/index'
import {useSelector, useDispatch} from 'react-redux';
import {GetMotos} from '../../redux/actions/actionCreators';
import {StackScreenProps} from '@react-navigation/stack';
import {Moto} from '../../models/moto'
import {
  Image,
  Text,
  View,
  ScrollView,
  StyleSheet,
  TouchableHighlight,
  SafeAreaView,
} from 'react-native';

interface Props extends StackScreenProps<any, any> {
  route: any
}

function MotosList({navigation, route}: Props) {
  const motos = useSelector((store:RootState) => store.motos);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(GetMotos());
  }, []);
  const {typeMoto} = route.params;
  const filteredMotos = motos.filter((moto: Moto) => (moto.tipo.toLowerCase() === typeMoto.toLowerCase()) || (moto.marca.toLowerCase() === typeMoto.toLowerCase()) || (moto.modelo.toLowerCase() === typeMoto.toLowerCase()) || (moto.cc === `${typeMoto} cc`));
  return (
    <SafeAreaView style={styles.background}>
      <ScrollView >
    <Image
    style={styles.logo}
    source={{
      uri: 'https://i.postimg.cc/FF2Y5GJd/logo-small.png',
    }}
  />
  <Text style={styles.typeMotoSelected}>{typeMoto.toUpperCase()}</Text>
        {filteredMotos.map((moto: Moto) => (
          <View style={styles.product} key={moto._id}>
            <TouchableHighlight
              onPress={() =>
                navigation.navigate('Detail', {motoSelected: moto._id})
              }>
              <Image
                testID = 'detailMoto'
                style={styles.foto}
                source={{
                  uri: moto.picture,
                }}
              />
            </TouchableHighlight>
            <View style={styles.features}>
              <Text style={styles.text}>Marca: {moto.marca}</Text>
              <Text style={styles.text}>Modelo: {moto.modelo}</Text>
              <Text style={styles.text}>Cilindrada: {moto.cc}</Text>
              <View style={styles.containerPrice}>
              <Text style={styles.price}>{moto.precio}€</Text>
              { moto.anteriorPrecio ? (
              <Text style={styles.realPrice}>{moto?.anteriorPrecio}€</Text>
              ):
              (
                <></>
                )}
                </View>
            </View>
          </View>
        ))}
      </ScrollView>
    </SafeAreaView>
  );
}
const styles = StyleSheet.create({
  foto: {
    width: 180,
    height: 100,
    borderRadius: 25,
    marginBottom: 50,
  },
  containerPrice:{
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  },
  typeMotoSelected: {
    color: 'white',
    fontSize: 18,
    marginBottom: 15,
    marginLeft: 20,
    borderBottomWidth: 1,
    width: 370,
    padding: 4,
    borderBottomColor: '#FFB800'
  },
  price: {
    fontSize: 20,
    color: '#FFB800'
  },
  realPrice: {
    color: 'red',
    fontSize: 20,
    textDecorationLine: "line-through"
  },
  logo: {
    width: 300,
    height: 42,
    marginTop: 15,
    marginBottom: 15,
    marginLeft: 10
  },
  background: {
    backgroundColor: 'black',
  },
  text: {
    color: 'white',
    fontSize: 18
  },
  backButton: {
    textAlign: 'center',
    marginBottom: 15,
  },
  product: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around'
  },
  features: {
    display: 'flex',
    flexDirection: 'column',
  },
});

export default MotosList;
