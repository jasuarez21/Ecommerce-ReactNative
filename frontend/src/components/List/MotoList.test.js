import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/actionCreators';
import { Provider } from 'react-redux';
import List from './MotosList';

describe('When invoked a List component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        motos: [{
              _id: '2',
              tipo: 'scooter',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com',
              anteriorPrecio: 300
          }
        ]
    });
    let route;
    let navigation;
    beforeEach(() => {
        jest.spyOn(actions, 'GetMotos').mockReturnValueOnce({ type: '' });
        route = {
            params: {
                typeMoto : 'scooter'
            }
        }
      });
      navigation = {
          navigate: jest.fn()
      };

    test('Should render a list of motos with type selected', () => {
        const list = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>)
        expect(list).toMatchSnapshot();
    })
    test('Then navigation.navigate detail is invoked', () => {
        const { getByTestId } = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>);
        const image = getByTestId('detailMoto');
        fireEvent.press(image);
        expect(navigation.navigate).toHaveBeenCalled();
      });
})

describe('When invoked a List component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        motos: [{
              _id: '2',
              tipo: 'scooter',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com',
              anteriorPrecio: 300
          }
        ]
    });
    let route;
    let navigation;
    beforeEach(() => {
        jest.spyOn(actions, 'GetMotos').mockReturnValueOnce({ type: '' });
        route = {
            params: {
                typeMoto : 'SR'
            }
        }
      });
      navigation = {
          navigate: jest.fn()
      };

    test('Should render a list of motos with type selected', () => {
        const list = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>)
        expect(list).toMatchSnapshot();
    })
    test('Then navigation.navigate detail is invoked', () => {
        const { getByTestId } = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>);
        const image = getByTestId('detailMoto');
        fireEvent.press(image);
        expect(navigation.navigate).toHaveBeenCalled();
      });
})

describe('When invoked a List component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        motos: [{
              _id: '2',
              tipo: 'scooter',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com',
              anteriorPrecio: 300
          }
        ]
    });
    let route;
    let navigation;
    beforeEach(() => {
        jest.spyOn(actions, 'GetMotos').mockReturnValueOnce({ type: '' });
        route = {
            params: {
                typeMoto : '50'
            }
        }
      });
      navigation = {
          navigate: jest.fn()
      };

    test('Should render a list of motos with type selected', () => {
        const list = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>)
        expect(list).toMatchSnapshot();
    })
    test('Then navigation.navigate detail is invoked', () => {
        const { getByTestId } = render(<Provider store={store}><List route={route} navigation={navigation} /></Provider>);
        const image = getByTestId('detailMoto');
        fireEvent.press(image);
        expect(navigation.navigate).toHaveBeenCalled();
      });
})