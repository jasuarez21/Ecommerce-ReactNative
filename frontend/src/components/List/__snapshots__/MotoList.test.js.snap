// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked a List component Should render a list of motos with type selected 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "black",
    }
  }
>
  <RCTScrollView>
    <View>
      <Image
        source={
          Object {
            "uri": "https://i.postimg.cc/FF2Y5GJd/logo-small.png",
          }
        }
        style={
          Object {
            "height": 42,
            "marginBottom": 15,
            "marginLeft": 10,
            "marginTop": 15,
            "width": 300,
          }
        }
      />
      <Text
        style={
          Object {
            "borderBottomColor": "#FFB800",
            "borderBottomWidth": 1,
            "color": "white",
            "fontSize": 18,
            "marginBottom": 15,
            "marginLeft": 20,
            "padding": 4,
            "width": 370,
          }
        }
      >
        SCOOTER
      </Text>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Image
            source={
              Object {
                "uri": "img.com",
              }
            }
            style={
              Object {
                "borderRadius": 25,
                "height": 100,
                "marginBottom": 50,
                "width": 180,
              }
            }
            testID="detailMoto"
          />
        </View>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Marca: 
            Aprilia
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Modelo: 
            SR
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Cilindrada: 
            50
          </Text>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-around",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFB800",
                  "fontSize": 20,
                }
              }
            >
              550
              €
            </Text>
            <Text
              style={
                Object {
                  "color": "red",
                  "fontSize": 20,
                  "textDecorationLine": "line-through",
                }
              }
            >
              300
              €
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`When invoked a List component Should render a list of motos with type selected 2`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "black",
    }
  }
>
  <RCTScrollView>
    <View>
      <Image
        source={
          Object {
            "uri": "https://i.postimg.cc/FF2Y5GJd/logo-small.png",
          }
        }
        style={
          Object {
            "height": 42,
            "marginBottom": 15,
            "marginLeft": 10,
            "marginTop": 15,
            "width": 300,
          }
        }
      />
      <Text
        style={
          Object {
            "borderBottomColor": "#FFB800",
            "borderBottomWidth": 1,
            "color": "white",
            "fontSize": 18,
            "marginBottom": 15,
            "marginLeft": 20,
            "padding": 4,
            "width": 370,
          }
        }
      >
        SR
      </Text>
      <View
        style={
          Object {
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <Image
            source={
              Object {
                "uri": "img.com",
              }
            }
            style={
              Object {
                "borderRadius": 25,
                "height": 100,
                "marginBottom": 50,
                "width": 180,
              }
            }
            testID="detailMoto"
          />
        </View>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Marca: 
            Aprilia
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Modelo: 
            SR
          </Text>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Cilindrada: 
            50
          </Text>
          <View
            style={
              Object {
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-around",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#FFB800",
                  "fontSize": 20,
                }
              }
            >
              550
              €
            </Text>
            <Text
              style={
                Object {
                  "color": "red",
                  "fontSize": 20,
                  "textDecorationLine": "line-through",
                }
              }
            >
              300
              €
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`When invoked a List component Should render a list of motos with type selected 3`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "black",
    }
  }
>
  <RCTScrollView>
    <View>
      <Image
        source={
          Object {
            "uri": "https://i.postimg.cc/FF2Y5GJd/logo-small.png",
          }
        }
        style={
          Object {
            "height": 42,
            "marginBottom": 15,
            "marginLeft": 10,
            "marginTop": 15,
            "width": 300,
          }
        }
      />
      <Text
        style={
          Object {
            "borderBottomColor": "#FFB800",
            "borderBottomWidth": 1,
            "color": "white",
            "fontSize": 18,
            "marginBottom": 15,
            "marginLeft": 20,
            "padding": 4,
            "width": 370,
          }
        }
      >
        50
      </Text>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
